src/main/java/net/minecraftforge/common/capabilities/LazyCapabilityProvider.java

```diff
@@ -0,0 +1,112 @@
+import net.minecraft.util.Direction;
+import net.minecraftforge.common.util.LazyOptional;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A capability provider whose caps are only attached lazily (when getCapability is called).
+ * When caps have not been initialized yet, capNBT is used the rest of the time for comparison and serialization.
+ * Used by forge for ItemStacks.
+ */
+public abstract class LazyCapabilityProvider<B extends LazyCapabilityProvider<B>> extends CapabilityProvider<B>
+{
+    private final CompoundNBT capNBT;
+    private volatile boolean capsInitialized = false; // must be volatile due for double-checked locking to be safe
+
+    protected LazyCapabilityProvider(Class<B> baseClass, @Nullable CompoundNBT capNBT)
+    {
+        super(baseClass);
+        this.capNBT = capNBT;
+    }
+
+    /**
+     * Called right before getCapability when necessary, init your capabilities there.
+     * DO NOT CALL getCapabilities(), or anything using it. deserializeCaps() is safe to use.
+     */
+    protected abstract void initializeCaps(@Nullable CompoundNBT capNBT);
+
+    private void initializeCapsIfNeeded()
+    {
+        // A note on thread safety: capsInitialized == true guarantees that capabilities are safe to access.
+        // Because we are not using a lock for reads, it is possible that in some cases capsInitialized might be false,
+        // but capabilities have already been initialized due to another thread.
+        // When that happens, we assume that using the immutable capNBT instead of the capabilities themselves is fine.
+        // This is equivalent to assuming both threads only read at the same time.
+        if (!this.capsInitialized)
+        {
+            synchronized (this)
+            {
+                if (!this.capsInitialized)
+                {
+                    initializeCaps(capNBT);
+                    this.capsInitialized = true;
+                }
+            }
+        }
+    }
+
+    @Nonnull
+    @Override
+    public <T> LazyOptional<T> getCapability(@Nonnull Capability<T> cap, @Nullable Direction side)
+    {
+        // Make sure caps are initialized before the query is resolved.
+        this.initializeCapsIfNeeded();
+        return super.getCapability(cap, side);
+    }
+
+    @Nullable
+    @Override
+    protected CapabilityDispatcher getCapabilities() {
+        // Detect calls that don't check for capsInitialized, or calls from external code.
+        if (!this.capsInitialized) throw new IllegalStateException("Capabilities are not yet initialized. This is a bug!");
+        return super.getCapabilities();
+    }
+
+    // Reimplementation of comparison that does not fire any capability event, relying on capNBT instead.
+    @Override
+    public boolean areCapsCompatible(@Nonnull CapabilityProvider<B> other)
+    {
+        LazyCapabilityProvider<B> o = (LazyCapabilityProvider<B>) other;
```

Comment:
I am not sure this is a safe cast to do as at least API wise another CapabilityProvider may be passed, so this should probably be instanced checked, and then for when not lazy compare it as if the other one is initialized already or maybe some other handling such as just returning false.

Reply:
I think this is fine with the `B extends LazyCapabilityProvider<B>` bound.

Reply:
No need for a cast anymore.